File: 104-maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:

        height = 0

        def dfs(root):

            nonlocal height
            if not root:
                return 0

            left = self.maxDepth(root.left)
            right = self.maxDepth(root.right)

            max_height = max(left, right)
            height = 1 + max_height

            return height

        return dfs(root)

File: 104-maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    int depth = 0;

    public int dfs(TreeNode root) {
        
        if (root == null) {
            return 0;
        }
        
        int left = dfs(root.left);
        int right = dfs(root.right);

        int max_height = Math.max(left, right);
        depth = 1 + max_height;

        return depth;
    }

    public int maxDepth(TreeNode root) {

        return dfs(root);
    }
}